contracts/contracts-foundry/src/kms/KmsRoot.sol

```diff
@@ -2,176 +2,82 @@
-    /// @param _owner Address of the contract owner
+    /// @param _admin Address of the contract admin
+    /// @param _approver Address of the image approver
+    /// @param _revoker Address of the image revoker
     /// @param _verifier Address of the RISC Zero verifier contract
-    /// @param _imageId Image ID for verification
-    /// @param _pcrs Initial PCR values
+    /// @param _guestId Guest ID of the RISC Zero program
     /// @param _rootKey Initial root key
-    /// @param _maxAge Maximum age allowed for attestation timestamps
+    /// @param _maxAgeMs Maximum age allowed for attestation timestamps
+    /// @param _imageId Image ID for verification
     constructor(
-        address _owner,
+        address _admin,
+        address _approver,
+        address _revoker,
         IRiscZeroVerifier _verifier,
-        bytes32 _imageId,
-        bytes memory _pcrs,
+        bytes32 _guestId,
         bytes memory _rootKey,
-        uint256 _maxAge
-    ) Ownable(_owner) {
-        MAX_AGE = _maxAge;
-
-        _updateVerifier(_verifier);
-        _updateImageId(_imageId);
-        _updatePcrs(_pcrs);
-        _updateRootKey(_rootKey);
+        uint256 _maxAgeMs,
+        bytes32 _imageId
+    ) RiscZeroVerifier(_verifier, _guestId, _rootKey, _maxAgeMs) VerifiedKeys(_imageId, DEFAULT_FAMILY) {
+        _grantRole(DEFAULT_ADMIN_ROLE, _admin);
+        _grantRole(APPROVER_ROLE, _approver);
+        _grantRole(REVOKER_ROLE, _revoker);
     }
 
-    /// @notice Updates the RISC Zero verifier contract address
-    /// @param _verifier New verifier contract address
-    function updateVerifier(IRiscZeroVerifier _verifier) external onlyOwner {
-        _updateVerifier(_verifier);
-    }
+    /// @notice Authorizes the admin to execute parameter updates
+    function _rzvAuthorizeUpdate() internal virtual override onlyRole(DEFAULT_ADMIN_ROLE) {}
 
-    /// @notice Updates the image ID used for verification
-    /// @param _imageId New image ID
-    function updateImageId(bytes32 _imageId) external onlyOwner {
-        _updateImageId(_imageId);
-    }
+    /// @notice Authorize the approver to approve images
+    function _vkAuthorizeApprove() internal virtual override onlyRole(APPROVER_ROLE) {}
 
-    /// @notice Updates the PCR values
-    /// @param _pcrs New PCR values
-    function updatePcrs(bytes calldata _pcrs) external onlyOwner {
-        _updatePcrs(_pcrs);
-    }
+    /// @notice Authorize the revoker to revoke images
+    function _vkAuthorizeRevoke() internal virtual override onlyRole(REVOKER_ROLE) {}
 
-    /// @notice Updates the root key
-    /// @param _rootKey New root key
-    function updateRootKey(bytes calldata _rootKey) external onlyOwner {
-        _updateRootKey(_rootKey);
+    /// @notice Tranform the public key into an address before storage
+    function _vkTransformPubkey(bytes memory _pubkey) internal virtual override returns (bytes32) {
+        require(_pubkey.length == 64, KmsRootLengthInvalid());
```

Comment:
Logic in `AttestationAuther` is slightly more efficient.

Reply:
done